mesos-rxjava-recordio/src/main/java/org/apache/mesos/rx/java/recordio/MessageStream.java

```diff
@@ -0,0 +1,110 @@
+package org.apache.mesos.rx.java.recordio;
+
+import com.google.common.base.Charsets;
+import io.netty.buffer.ByteBuf;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.BufferOverflowException;
+import java.nio.ByteBuffer;
+import java.nio.channels.Channels;
+import java.nio.channels.ReadableByteChannel;
+
+final class MessageStream {
+
+    @NotNull
+    private static final Logger LOGGER = LoggerFactory.getLogger(MessageStream.class);
+
+    /** Maximum number of base-10 digits in a uint64 */
+    private static final int MESSAGE_SIZE_MAX_LENGTH = 20;
+
+    @NotNull
+    private final InputStream source;
+
+    @NotNull
+    private final ReadableByteChannel channel;
+
+    /**
+     * The message size from the stream is provided as a base-10 String representation of an
+     * unsigned 64 bit integer (uint64). Since there is a possibility (since the spec isn't
+     * formal on this, and the HTTP chunked Transfer-Encoding applied to the data stream can
+     * allow chunks to be any size) this field functions as the bytes that have been read
+     * since the end of the last message. When the next '\n' is encountered in the byte
+     * stream, these bytes are converted to a UTF-8 String. This string representation is
+     * then read as a {@code long} using {@link Long#valueOf(String, int)}.
+     */
+    @NotNull
+    private final ByteBuffer messageSize = ByteBuffer.allocate(MESSAGE_SIZE_MAX_LENGTH);
+
+    /** The buffer for the current message being read from the stream. */
+    @NotNull
+    private ByteBuffer message = ByteBuffer.allocate(0);
+
+
+    public MessageStream(@NotNull final InputStream source) {
+        this.source = source;
+        this.channel = Channels.newChannel(source);
```

Comment:
Using both `source` and `channel` is potentially confusing since they read from the same byte sequence. But we need `source` to read single bytes (for parsing the message size) and `channel` to read into `ByteBuffer message` in bulk (for parsing the message).

The alternatives are using either one or the other exclusively; here are the drawbacks of each approach:
- Only `InputStream source`: we'd need to manually read into `ByteBuffer message` using `source.read(message.array(), message.arrayOffset(), message.limit())` and then manually adjust the offset of `message`.
- Only `ReadableByteChannel channel`: we'd need to keep a single-byte `ByteBuffer` around to read each byte of the message size using `ReadableByteChannel.read(ByteBuffer)`.